### DISCUSSION TITLE
Saving participant code/ID from a survey-text as an output data column of audio-button-response experiment

### DISCUSSION THREAD

**Discussion by tatulhum** [jspsych/jsPsych#1163] [Q&A] [answered]
I would like participants to be able to input their names into a survey-text and for this to be made a column of the data output for an audio-button-response. Currently, I have been able to use the example here: https://www.jspsych.org/overview/data/ to get a random ID appended:

`var subject_id = jsPsych.randomization.randomID(15);`

`jsPsych.data.addProperties({ subject: subject_id});`

I do not want the ID to be random, though, but something I input myself. I have a survey set up, but am not sure how to get the response to record. I know that "undefined" is not right, but am not sure what is. Can someone tell me what I should put under "on_finish"  in the survey-text and what I should change "var subject_id" to in order to link the two?

`var contact = {`
   ` type: 'survey-text-custom',`
  `  preamble: "<h2>Please enter your name</h2>",`
  `  json_label: 'Contact',`
 `   questions: [
        {prompt: "Your name:", name: 'Name'},
    ],`
  `  //on_load: function () {
      //return undefined;
    //},`
   ` on_finish: function () { 
      return undefined;
    }
};`

`var subject_id = jsPsych.data.get().filter({trial_type: 'survey-text-custom', name: 'Name'})`

Added `jsPsych.data.addProperties({ subject: subject_id});` in the audio-button-response section:

 ` for(var i=0; i<sound_stim_shuffled.length; i++) {`
 `    index = i + 1`
`     new_screen = {`
       ` type: 'audio-button-response',`
  `      stimulus: 'sound/' + sound_stim_shuffled[i].sound,`
    `    choices: [first_word,
                  second_word,
                  third_word],`
  `      button_html:['<img src="'+word_set.first_word+'" width="250px" />',
                    '<img src="'+word_set.second_word+'" width="250px" />',
                    '<img src="'+word_set.third_word+'" width="250px" />',
                    '<img src="img/sound1.jpg" width="150px" />',],`
   `     //margin_horizontal: '50px',`
`prompt: '<table>' +
        '<tr style="line-height:200%"><th>&nbsp;</tr>' +
        '<tr style="font-size:200%"><th><strong> Please click what you heard.</strong></th></tr>' +
        '<tr style="line-height:200%"><th>&nbsp;</tr>' +
        '<tr style="font-size:150%"><td align="right">' + index + '/' + sound_stim_shuffled.length + '</td></tr></table>',
        post_trial_gap: 500,            // interval between trials(ms)`
      `  data:{main_sound:sound_stim[0].sound, corrans: sound_stim_shuffled[i].corrans},`
     `   on_finish: function(data) {     // Check correctness`
         ` // debugger;`
      `     var correct = false;`
`           if(data.corrans == data.button_pressed) {
               correct = true;
           }`
      `  jsPsych.data.addDataToLastTrial({correct: correct});`
      `  jsPsych.data.addProperties({subject: subject_id,});`
 `       }`
`     }`
`     new_qst_trial.actual_screen.push(make_loop_node(new_screen));`
  ` }`

**Comment by tatulhum** [comment]
Or alternatively, can anyone suggest other ways of recording a participant ID that is input by the participant themself? I'm new to this, so maybe I am missing how people normally do this. Any examples or advice would be much appreciated!

**Comment by wenbo-lin** [comment, marked as answer]
There is probably a better way to do this. I just quickly wrote this example.  The goal is to grab the data from the survey-text plugin. In this case, you can find them with `data.responses`. 
```
var subjectID_page = {
    type: 'survey-text',
    questions: [
      {prompt: 'Enter your subject ID', columns: 50, required: true, name: 'subject_id'}
    ],
	on_finish:function(data){
	let userResponses = JSON.parse(data.responses);
	
	jsPsych.data.addProperties({subject: userResponses.subject_id});
	}
  };

 var age_question = {
    type: 'survey-text',
    questions: [
      {prompt: 'Enter your age', columns: 3, required: true, name: 'age'}
    ],
};

  jsPsych.init({
    timeline: [subjectID_page, age_question],
    on_finish: function() { jsPsych.data.displayData(); }
  });
```

**Reply by tatulhum** [reply]
Oh, thanks, I just saw your reply! I will try this

**Reply by tatulhum** [reply]
This works in that subject_id gets created as a data column, but the participant code entered does not save to it. My timeline is a variable, but I've added the survey-text questions to it and that has been fine. However, I'm using a script that has a different function under jsPsych.init than `on_finish: function() { jsPsych.data.displayData(); }`. I'm not sure how to incorporate the two. The function filters out some unneeded parts of the audio-button-response data, but possibly it's also filtering out the survey-text too?

`jsPsych.init({`
    `timeline: timeline,`
   ` show_progress_bar: true,`
   ` on_finish:function(data) {`
       ` data = jsPsych.data.get().filter({trial_type:'audio-button-response'}).ignore('trial_index').ignore('time_elapsed').ignore('internal_node_id')`
        `saveData('dene_'+stamp()+'_result.csv', data.csv());`
       ` finish();`
  `  }`
`});`

The functions in jsPsych.init:

`function saveData(filename, filedata){`
`    $.ajax({`
        `type:'post',`
      `  cache: false,`
      `  url: 'save_data.php',`
        `data: {filename: filename, filedata: filedata}`
   ` });`
`};`

`function finish() {`
    `DOM_target = document.querySelector('#jspsych-content');`
    `DOM_target.innerHTML ='<p style="font-size:200%;"><strong>Thank you for participating</strong></p>' ;`
    `jsPsych.pluginAPI.cancelAllKeyboardResponses();`
    `jsPsych.pluginAPI.clearAllTimeouts();`
`};`

**Reply by wenbo-lin** [reply]
I think you did filter out the data generated from the survey-text trial/plugin with the following line:

 `data = jsPsych.data.get().filter({trial_type:'audio-button-response'})`. This line will only grab data generated from the audio-button-response plugin. Is there a specific reason why you're using filter here?

If you simply don't want the internal_node_id column, you can do the following: `jsPsych.data.get().ignore('internal_node_id')`.

**Reply by tatulhum** [reply]
The filter was to remove instruction trials, as well as some unneeded columns. If I get rid of the filter, the inputted subject_id saves as a response for the survey-text, but still does not save under addProperties. subject remains blank:

![image](https://user-images.githubusercontent.com/67024217/97934284-3db1aa00-1d43-11eb-9ec6-f072f8f589c9.png)

I'm not sure why that would be?


`var subjectID_page = {`
  `  type: 'survey-text',`
  `  questions: [`
   `   {prompt: 'Enter your subject ID', columns: 50, required: true, name: 'subject_id'}`
  `  ],`
`	on_finish:function(data){`
	`let userResponses = JSON.parse(data.responses);`
	`jsPsych.data.addProperties({subject: userResponses.subject_id});`
	`}`
 ` };`

`var timeline = [instruction_trial, subjectID_page];`
`for (var i = 0; i < qst_trial.length; i++) {`
  `timeline = timeline.concat(qst_trial[i].pre_trial_instructions);`
 ` timeline = timeline.concat(qst_trial[i].actual_screen);`
`};`

`jsPsych.init({`
   ` timeline: timeline,`
    `show_progress_bar: true,`
   ` on_finish: function(data) {`
        `data = jsPsych.data.get()`
        `saveData('dene_'+stamp()+'_result.csv', data.csv());`
       ` finish();`
  `  }`
`});`

**Reply by wenbo-lin** [reply]
That's odd. I don't have the full version of your codes, so I just used my earlier example.

```
  var subjectID_page = {
    type: 'survey-text',
    questions: [
      {prompt: 'Enter your subject ID', columns: 50, required: true, name: 'subject_id'}
    ],
	on_finish:function(data){
	let userResponses = JSON.parse(data.responses);
	
	jsPsych.data.addProperties({subject: userResponses.subject_id});
	}
  };

 var age_question = {
    type: 'survey-text',
    questions: [
      {prompt: 'Enter your age', columns: 3, required: true, name: 'age'}
    ],
};


var timeline = [subjectID_page, age_question];

  jsPsych.init({
    timeline: timeline,
    on_finish: function() { 
	jsPsych.data.get().localSave('csv','mydata.csv');
	}
```

![image](https://user-images.githubusercontent.com/5298587/97936680-e959fa80-1d41-11eb-9382-81b06b004b9a.png)

**Reply by wenbo-lin** [reply]
I don't know if you are using this portion that you posted earlier.

```
for(var i=0; i<sound_stim_shuffled.length; i++) {
index = i + 1
new_screen = {
type: 'audio-button-response',
stimulus: 'sound/' + sound_stim_shuffled[i].sound,
choices: [first_word, second_word, third_word],
button_html:['<img src="'+word_set.first_word+'" width="250px" />', '<img src="'+word_set.second_word+'" width="250px" />', '<img src="'+word_set.third_word+'" width="250px" />', '<img src="img/sound1.jpg" width="150px" />',],
//margin_horizontal: '50px',
prompt: '<table>' + '<tr style="line-height:200%"><th>&nbsp;</tr>' + '<tr style="font-size:200%"><th><strong> Please click what you heard.</strong></th></tr>' + '<tr style="line-height:200%"><th>&nbsp;</tr>' + '<tr style="font-size:150%"><td align="right">' + index + '/' + sound_stim_shuffled.length + '</td></tr></table>', post_trial_gap: 500, // interval between trials(ms)
data:{main_sound:sound_stim[0].sound, corrans: sound_stim_shuffled[i].corrans},
on_finish: function(data) { // Check correctness
// debugger;
var correct = false;
if(data.corrans == data.button_pressed) { correct = true; }
jsPsych.data.addDataToLastTrial({correct: correct});
jsPsych.data.addProperties({subject: subject_id,});
}
}
new_qst_trial.actual_screen.push(make_loop_node(new_screen));
}
```

If so, you might be overwriting the subject column with the line `jsPsych.data.addProperties({subject: subject_id,});` that is under the line `jsPsych.data.addDataToLastTrial({correct: correct});`

**Reply by tatulhum** [reply]
I had taken that out. Here is the full version of the code: [https://github.com/tatulhum/listening-experiment/blob/main/experiment_vDeneTest.html](url). If I do localSave on the full version, the same thing happens where the result is an empty column. Is there anything in the full version that may be causing this?

**Reply by wenbo-lin** [reply]
Because I don't have the audio files, I just used one random audio file and made it into one audio trial.
It works fine.
![image](https://user-images.githubusercontent.com/5298587/97946435-085f8900-1d50-11eb-8d26-42f8972c38ef.png)

You might want to check the scripts you're loading in the `<head></head>` tag. For example, what is the difference between the survey-text plugin and the survey-text-custom plugin? If you don't need survey-text-custom, then remove it.  For my example, I only loaded the following plugins.

```
        <script src="../plugins/jspsych-survey-text.js"></script>
	<script src="../plugins/jspsych-instructions.js"></script>
	<script src="../plugins/jspsych-audio-keyboard-response.js"></script>
	<script src="../plugins/jspsych-audio-button-response.js"></script>
```

This was what I used to generate the screenshot above: 
```
<!DOCTYPE html>
<html>
  <head>

    <script src="../jspsych.js"></script>
    <script src="../plugins/jspsych-survey-text.js"></script>
	<script src="../plugins/jspsych-instructions.js"></script>
	<script src="../plugins/jspsych-audio-keyboard-response.js"></script>
	<script src="../plugins/jspsych-audio-button-response.js"></script>
    <link rel="stylesheet" href="../css/jspsych.css"></link>
  </head>
  <body></body>
  <script>
  
 /**********************************************************************************************************************/
/******************************** Instruction trials ******************************************************************/
/**********************************************************************************************************************/
var instStim = '<p style="font-size:200%;"><strong> Welcome.</strong></p><br>' +
               '<p style="font-size:150%;"> Your task will be to listen to each recording and select the word that you hear.</p>';

var instruction_trial = {
    type: 'instructions',
    pages: [ instStim ] ,
    show_clickable_nav: true
};

var subjectID_page = {
    type: 'survey-text',
    questions: [
      {prompt: 'Enter your subject ID', columns: 50, required: true, name: 'subject_id'}
    ],
	on_finish:function(data){
	let userResponses = JSON.parse(data.responses);

	jsPsych.data.addProperties({subject: userResponses.subject_id});
	}
  };

 var age_question = {
    type: 'survey-text',
    questions: [
      {prompt: 'Enter your age', columns: 3, required: true, name: 'age'}
    ],
};


var gender = {
    type: 'gender-info',
    preamble: "<h2>Gender</h2>",
    json_label: 'Gender',
    on_load: function ()
    {
        validate_gender();
        form = document.getElementById("jspsych-content");
        toggle_listeners(form, true, validate_gender);
    },
    on_finish: function () {
        toggle_listeners(form, false, validate_gender);
    }
};

/**********************************************************************************************************************/
/******************************** Listening Trials ********************************************************************/
/**********************************************************************************************************************/



function global_loop_function(data){
  var data = jsPsych.data.get().last(1).values()[0];
    if(data.button_pressed == 3){
        return true;
    } else {
        return false;
    };
}

function make_loop_node(screen){
  screen.choices.push('img/sound1.jpg');
  var looped_node = {
      timeline: [screen],
      loop_function: global_loop_function,
  }
  return looped_node;
}

var word_sets = [
{
  first_word: 'img/d_really.png',
  second_word: 'img/t_lake.png',
  third_word: 'img/t_ej_bark.png',
  sound_stim: [
  {sound:'bird.wav', corrans:1},
                   ],
},
];

function new_set(word_set){
  var first_word = word_set.first_word,
      second_word = word_set.second_word,
      third_word = word_set.third_word,
      sound_stim = word_set.sound_stim,
      new_qst_trial = {};

  new_qst_trial.pre_trial_instructions = {
      type: 'instructions',
      pages: [ '<p style="font-size:200%;"><strong> Listen to each recording and select the image representing the word that you hear. </strong></p><br>' +
                '<p style="font-size:200%;">Here are the three words you will be choosing between:</p><br>' +
                '<img src="'+word_set.first_word+'" width="200"/>'+
                '<img src="'+word_set.second_word+'" width="200"/>'+
                '<img src="'+word_set.third_word+'" width="200"/>' +
                '<p style="font-size:150%;"> Are you ready? The first word will play when you press "Next".</p>' ] ,
      show_clickable_nav: true
  }

  sound_stim_repeated = jsPsych.randomization.repeat(sound_stim, 1);   //randomize and repeat x times
  sound_stim_shuffled = jsPsych.randomization.shuffleNoRepeats(sound_stim_repeated, function(a,b) {return a.sound === b.sound});

  new_qst_trial.actual_screen = [];

  for(var i=0; i<sound_stim_shuffled.length; i++) {
     index = i + 1
     new_screen = {
        type: 'audio-button-response',
        stimulus: 'sound/' + sound_stim_shuffled[i].sound,
        choices: [first_word,
                  second_word,
                  third_word],
        button_html:['<img src="'+word_set.first_word+'" width="250px" />',
                    '<img src="'+word_set.second_word+'" width="250px" />',
                    '<img src="'+word_set.third_word+'" width="250px" />',
                    '<img src="img/sound1.jpg" width="150px" />',],
        //margin_horizontal: '50px',
prompt: '<table>' +
        '<tr style="line-height:200%"><th>&nbsp;</tr>' +
        '<tr style="font-size:200%"><th><strong> Please click what you heard.</strong></th></tr>' +
        '<tr style="line-height:200%"><th>&nbsp;</tr>' +
        '<tr style="font-size:150%"><td align="right">' + index + '/' + sound_stim_shuffled.length + '</td></tr></table>',
        post_trial_gap: 500,            // interval between trials(ms)
        data:{main_sound:sound_stim[0].sound, corrans: sound_stim_shuffled[i].corrans},
        on_finish: function(data) {     // Check correctness
          // debugger;
           var correct = false;
           if(data.corrans == data.button_pressed) {
               correct = true;
           }
        jsPsych.data.addDataToLastTrial({correct: correct});
      //  jsPsych.data.addProperties({subject: subject_id,});
        }
     }
     new_qst_trial.actual_screen.push(make_loop_node(new_screen));
   }

  return new_qst_trial;

}



var qst_trial = [];

for (var i = 0; i < word_sets.length; i++) {
  // debugger;
  qst_trial.push(new_set(word_sets[i]));
}



/**********************************************************************************************************************/
/******************************** User defined function ***************************************************************/
/**********************************************************************************************************************/

//save data function
function saveData(filename, filedata){
    $.ajax({
        type:'post',
        cache: false,
        url: 'save_data.php',
        data: {filename: filename, filedata: filedata}
    });
};

//time stamp function, format YYYYMMDDhhmmss
function stamp(){
    var dt = new Date();
    var year = dt.getFullYear();
    var month = ((dt.getMonth()+1)<10?'0':'') + (dt.getMonth()+1);
    var day = (dt.getDate()<10?'0':'') + dt.getDate();
    var hour = (dt.getHours()<10?'0':'') + dt.getHours();
    var minute = (dt.getMinutes()<10?'0':'') + dt.getMinutes();
    var second = (dt.getSeconds()<10?'0':'') + dt.getSeconds();
    var stamp = year +''+ month +''+ day +''+ hour +''+ minute +''+ second ;
    return stamp;
};


//Thank you message
function finish() {
    DOM_target = document.querySelector('#jspsych-content');
    DOM_target.innerHTML ='<p style="font-size:200%;"><strong>Thank you for participating</strong></p>' ;
    jsPsych.pluginAPI.cancelAllKeyboardResponses();
    jsPsych.pluginAPI.clearAllTimeouts();
};

/**********************************************************************************************************************/
/******************************** Make timeline ***********************************************************************/
/**********************************************************************************************************************/

var timeline = [instruction_trial, subjectID_page];
for (var i = 0; i < qst_trial.length; i++) {
  timeline = timeline.concat(qst_trial[i].pre_trial_instructions);
  timeline = timeline.concat(qst_trial[i].actual_screen);
};
/**********************************************************************************************************************/
/******************************** Run jsPsych *************************************************************************/
/**********************************************************************************************************************/

//jsPsych.init({
//    timeline: timeline,
//    show_progress_bar: true,
//    on_finish:function(data) {
//        data = jsPsych.data.get().filter({trial_type:'audio-button-response'}).ignore('trial_index').ignore('time_elapsed').ignore('internal_node_id')
//        saveData('dene_'+stamp()+'_result.csv', data.csv());
        //finish();
//});

jsPsych.init({
    timeline: timeline,
    show_progress_bar: true,
    on_finish: function(data) {
        jsPsych.data.get().localSave('csv','mydata.csv');
        finish();
    }
});

  </script>
</html>


```
Alternatively, you could try changing `jsPsych.data.addProperties({subject: userResponses.subject_id});` to `jsPsych.data.addProperties({subjectID: userResponses.subject_id});` if your data file contains both a "subject" column and a "subjectID" column, it means that somewhere in your code (including the code in your external files) you have another instance of `jsPsych.data.addProperties({subject:subject_id})`

**Comment by tatulhum** [comment]
My file was exactly the same as yours, so I tried updating from jsPsych 6.0.3 to 6.1, and after doing that I can now get the same output as you with the subject ID copied into the subject column when saving using `jsPsych.data.get().localSave('csv','mydata.csv');`. However, if I try to save to the server the way I was doing previously, the experiment will no longer load. 

This is how I was saving:

`jsPsych.init({`
    `  timeline: timeline,`
     ` show_progress_bar: true,`
      `on_finish: function(data) {`
          `data = jsPsych.data.get()`
          `saveData('dene_'+stamp()+'_result.csv', data.csv());`
         ` finish();`
`});`

`function saveData(filename, filedata){`
   ` $.ajax({`
      `  type:'post',`
        `cache: false,`
       ` url: 'save_data.php',`
       ` data: {filename: filename, filedata: filedata}`
   ` });`
`};`

The save_data.php that is used looks like the following:

`<?php`
 `   // the $_POST[] array will contain the passed in filename and data`
`    // the directory "data" is writable by the server (chmod 777)`
 `   $filename = "data/".$_POST['filename'];`
 `   $data = $_POST['filedata'];`
 `   // write the file to disk`
`    file_put_contents($filename, $data);`
`?>`

Is anything in these incompatible with v. 6.1 in terms of their syntax or something else which may have changed that I'm not aware of?

**Reply by wenbo-lin** [reply]
Check the browser console for error messages. Some of the file urls (file paths) may not be linked correctly.

**Reply by tatulhum** [reply]
I'm still mostly a beginner at this, so I'd forgotten about looking there for error messages. As it turns out, I was missing a bracket and also needed to put back the  <script src="jquery/jquery-3.3.1.min.js"></script> line. It is working now. Thank you so much for your help!
